tests/lib.rs

```diff
@@ -38,7 +44,24 @@ pub fn test_markdown_html(input: &str, output: &str, smart_punct: bool, metadata
     let p = Parser::new_ext(input, opts);
     pulldown_cmark::html::push_html(&mut s, p);
 
-    assert_eq!(normalize_html(output), normalize_html(&s));
+    // normalizing the HTML using html5ever may hide actual errors
+    // assert_eq!(normalize_html(output), normalize_html(&s));
+    assert_eq!(html_standardize(output), html_standardize(&s));
+}
+
+fn html_standardize(s: &str) -> String {
+    s.trim()
+        .replace("<br>", "<br />")
+        .replace("<br/>", "<br />")
+        .replace("<hr>", "<hr />")
+        .replace("<hr/>", "<hr />")
+        .replace("\n    ", "\n")
+        .replace("  \n", "\n")
+        .replace(" \n", "\n")
+        .replace("\n  ", "\n")
+        .replace("\n ", "\n")
```

Comment:
On closer inspection, I'm also not sure about these. Extra spaces at the end of lines are probably not something we want to keep in tests for no good reason.

- `  \n` -> ` \n` doesn't affect any tests. It could maybe be removed.
- ` \n` -> `\n` might actually highlight a point where we go against the spec.
  In 6.8 Soft line breaks:
  > Spaces at the end of the line and beginning of the next line are removed:
  
  But we preserve the space (there's one after a):
  ```
  $ target/debug/pulldown-cmark -e
  a 
  b
  0..5: Start(Paragraph)
  0..2: Text(Borrowed("a "))
  2..3: SoftBreak
  3..4: Text(Borrowed("b"))
  0..5: End(Paragraph)
  EOF
  ```
  And in 4.8 Paragraphs:
  > The paragraph’s raw content is formed by concatenating the lines and removing initial and final spaces or tabs.
  
  But we don't do that:
  ```
  $ target/debug/pulldown-cmark -e
  a 
  0..3: Start(Paragraph)
  0..2: Text(Borrowed("a "))
  0..3: End(Paragraph)
  EOF
  ```
  (these spaces are also preserved in the HTML output)
- `\n  ` -> `\n` affects one test from the CommonMark spec, [`spec_test_191`](https://spec.commonmark.org/0.30/#example-191). Though it's not explicitly mentioned, [example 183](https://spec.commonmark.org/0.30/#example-183) suggests that indentation on the first line of block HTML should be preserved, as commonmark.js does.
- `\n ` -> `\n` also doesn't affect any tests.

Reply:
Given these, I'm also a little skeptical that `.trim()` is useful. Maybe just `.strip_suffix('\n')`, shouldn't that be enough?

Reply:
Hm... I will dig into this, but all tests should pass so those issues must be resolved before changing this.

Thanks!

Reply:
Filed issue #715 on this